File: di-ipv-cimit-stub/lib/src/test/java/uk/gov/di/ipv/core/library/service/CimitStubItemServiceTest.java
package uk.gov.di.ipv.core.library.service;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import uk.gov.di.ipv.core.library.persistence.DataStore;
import uk.gov.di.ipv.core.library.persistence.items.CimitStubItem;

import java.time.Instant;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static uk.gov.di.ipv.core.library.config.ConfigurationVariable.CIMIT_STUB_TTL;

@ExtendWith(MockitoExtension.class)
class CimitStubItemServiceTest {

    private static final String USER_ID = "user-id-1";
    private static final String DB_TTL = "1800";
    @Captor private ArgumentCaptor<CimitStubItem> cimitStubItemArgumentCaptor;
    @Mock private ConfigService mockConfigService;
    @Mock private DataStore<CimitStubItem> mockDataStore;
    @InjectMocks private CimitStubItemService cimitStubItemService;

    @Test
    void shouldReturnCredentialIssuersFromDataStoreForSpecificUserId() {
        String ciCode = "V03";
        List<CimitStubItem> cimitStubItems =
                List.of(
                        CimitStubItem.builder()
                                .userId(USER_ID)
                                .contraIndicatorCode(ciCode)
                                .build());

        when(mockDataStore.getItems(USER_ID)).thenReturn(cimitStubItems);

        var result = cimitStubItemService.getCIsForUserId(USER_ID);

        assertTrue(
                result.stream()
                        .map(CimitStubItem::getContraIndicatorCode)
                        .anyMatch(ciCode::equals));
    }

    @Test
    void shouldCreateCimitStubItem() {
        String ciCode = "V03";
        List<String> mitigations = List.of("V01", "V03");
        Instant issuanceDate = Instant.now();
        List<String> issuers = List.of("https://address-cri.stubs.account.gov.uk");
        String docId = "some/document/id";

        cimitStubItemService.persistCimitStubItem(
                CimitStubItem.builder()
                        .userId(USER_ID)
                        .contraIndicatorCode(ciCode)
                        .issuers(issuers)
                        .issuanceDate(issuanceDate)
                        .mitigations(mitigations)
                        .documentIdentifier(docId)
                        .build());

        verify(mockDataStore).create(cimitStubItemArgumentCaptor.capture(), any());

        CimitStubItem capturedItem = cimitStubItemArgumentCaptor.getValue();

        assertEquals(USER_ID, capturedItem.getUserId());
        assertEquals(ciCode, capturedItem.getContraIndicatorCode());
        assertEquals(issuers, capturedItem.getIssuers());
        assertEquals(issuanceDate, capturedItem.getIssuanceDate());
        assertEquals(mitigations, capturedItem.getMitigations());
        assertEquals(docId, capturedItem.getDocumentIdentifier());
    }

    @Test
    void persistCimitStubItemShouldCreateAnItem() {
        CimitStubItem itemToPersist = CimitStubItem.builder().userId("some-user").build();

        cimitStubItemService.persistCimitStubItem(itemToPersist);

        verify(mockDataStore).create(cimitStubItemArgumentCaptor.capture(), eq(CIMIT_STUB_TTL));
        assertEquals(itemToPersist, cimitStubItemArgumentCaptor.getValue());
    }

    @Test
    void shouldUpdateCimitStubItem() {
        CimitStubItem cimitStubItem =
                CimitStubItem.builder()
                        .userId(USER_ID)
                        .contraIndicatorCode("D01")
                        .mitigations(List.of("V01", "V03"))
                        .issuanceDate(Instant.now())
                        .build();
        when(mockConfigService.getSsmParameter(CIMIT_STUB_TTL)).thenReturn(DB_TTL);
        cimitStubItemService.updateCimitStubItem(cimitStubItem);
        verify(mockDataStore, times(1)).update(any());
    }

    @Test
    void shouldDeleteCimitStubItem() {
        String ciCode = "D02";
        cimitStubItemService.deleteCimitStubItem(USER_ID, ciCode);
        verify(mockDataStore, times(1)).delete(any(), any());
    }
}

File: di-ipv-cimit-stub-ts/lambdas/test/internal-api/common/cimitStubItemService.test.ts
jest.mock("../../../src/common/configService", () => ({
  getCimitStubTableName: jest.fn().mockReturnValue("mock-table"),
  getCimitStubTtl: jest.fn().mockResolvedValue(1800),
}));

jest.mock("../../../src/clients/dynamoDBClient", () => ({
  dynamoDBClient: {
    query: jest.fn(),
    putItem: jest.fn(),
    deleteItem: jest.fn(),
  },
}));

import { dynamoDBClient } from "../../../src/clients/dynamoDBClient";
import * as cimitStubItemService from "../../../src/common/cimitStubItemService";
import { CimitStubItem } from "../../../src/common/contraIndicatorTypes";
import { marshall } from "@aws-sdk/util-dynamodb";

const USER_ID = "user-id-1";

beforeEach(() => {
  jest.clearAllMocks();
});

describe("getCIsForUserId", () => {
  it("should return CIs from DynamoDB for specific userId", async () => {
    const ciCode = "V03";
    const mockItems = [
      {
        userId: { S: USER_ID },
        contraIndicatorCode: { S: ciCode },
        sortKey: { S: `${ciCode}#2023-08-17T10:20:53.000Z` },
      },
    ];

    (dynamoDBClient.query as jest.Mock).mockResolvedValue({
      Items: mockItems,
    });

    const result = await cimitStubItemService.getCIsForUserId(USER_ID);

    expect(dynamoDBClient.query).toHaveBeenCalledWith({
      TableName: "mock-table",
      KeyConditionExpression: "userId = :userId",
      ExpressionAttributeValues: marshall({ ":userId": USER_ID }),
    });
    expect(result).toHaveLength(1);
    expect(result[0].contraIndicatorCode).toBe(ciCode);
  });
});

describe("getCiForUserId", () => {
  it("should return CIs matching sortKey prefix", async () => {
    const ciCode = "D02";
    const mockItems = [
      {
        userId: { S: USER_ID },
        contraIndicatorCode: { S: ciCode },
        sortKey: { S: `${ciCode}#2023-08-17T10:20:53.000Z` },
      },
    ];

    (dynamoDBClient.query as jest.Mock).mockResolvedValue({
      Items: mockItems,
    });

    const result = await cimitStubItemService.getCiForUserId(USER_ID, ciCode);

    expect(dynamoDBClient.query).toHaveBeenCalledWith({
      TableName: "mock-table",
      KeyConditionExpression:
        "userId = :userId AND begins_with(sortKey, :prefix)",
      ExpressionAttributeValues: marshall({
        ":userId": USER_ID,
        ":prefix": ciCode,
      }),
    });
    expect(result).toHaveLength(1);
  });
});

describe("persistCimitStubItem", () => {
  it("should create CimitStubItem with TTL", async () => {
    const ciCode = "V03";
    const mitigations = ["V01", "V03"];
    const issuanceDate = "2023-08-17T10:20:53.000Z";
    const issuer = "https://address-cri.stubs.account.gov.uk";
    const docId = "some/document/id";

    const item: CimitStubItem = {
      userId: USER_ID,
      contraIndicatorCode: ciCode,
      issuer,
      issuanceDate,
      mitigations,
      document: docId,
      sortKey: "",
      ttl: 0,
      txn: "",
    };

    await cimitStubItemService.persistCimitStubItem(item);

    expect(dynamoDBClient.putItem).toHaveBeenCalledTimes(1);
    const callArgs = (dynamoDBClient.putItem as jest.Mock).mock.calls[0][0];
    expect(callArgs.TableName).toBe("mock-table");
    expect(callArgs.Item).toBeDefined();
    expect(callArgs.Item.ttl.N).not.toBe("0");
    expect(callArgs.Item.sortKey.S).toBe(ciCode + "#" + issuanceDate);
  });
});

describe("updateCimitStubItem", () => {
  it("should update CimitStubItem with new TTL", async () => {
    const item: CimitStubItem = {
      userId: USER_ID,
      contraIndicatorCode: "D01",
      mitigations: ["V01", "V03"],
      issuanceDate: "2023-08-17T10:20:53.000Z",
      issuer: "https://test.example.com",
      document: "",
      sortKey: "",
      ttl: 0,
      txn: "",
    };

    await cimitStubItemService.updateCimitStubItem(item);

    expect(dynamoDBClient.putItem).toHaveBeenCalledTimes(1);
    const callArgs = (dynamoDBClient.putItem as jest.Mock).mock.calls[0][0];
    expect(callArgs.TableName).toBe("mock-table");
    expect(callArgs.Item).toBeDefined();
    expect(callArgs.Item.ttl.N).not.toBe("0");
    expect(callArgs.Item.sortKey.S).toBe("D01#2023-08-17T10:20:53.000Z");
  });
});

describe("deleteCimitStubItem", () => {
  it("should delete CimitStubItem by userId and sortKey", async () => {
    const sortKey = "D02#2023-08-17T10:20:53.000Z";

    await cimitStubItemService.deleteCimitStubItem(USER_ID, sortKey);

    expect(dynamoDBClient.deleteItem).toHaveBeenCalledWith({
      TableName: "mock-table",
      Key: marshall({
        userId: USER_ID,
        sortKey,
      }),
    });
  });
});
